[None][refactor] Improve lookahead decoding interfaces

Funatiq opened the pull request:
## Description

- Unify lookahead decoding functions with decoder setup.
- Remove `LlmRequest` from `DecoderState` and `GptDecoderBatched` interfaces.
- Remove the binding for fusing the sampling configs.

Please see commit messages for details.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Consolidated per-request sampling settings into a single fused sampling configuration and unified batch-slot handling to simplify decoder request flow.
  * Narrowed decoder setup to final context chunks and centralized lookahead-disabling logic, reducing redundant parameters and clarifying stream synchronization.
* **Chores**
  * Removed legacy helper bindings for combining sampling configs from the Python/C++ module surfaces.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Funatiq committed:
refactor: Simplify disableLookahead method in DecoderState

- Removed the RequestVector parameter from the disableLookahead method in DecoderState, streamlining its interface.
- Updated calls to disableLookahead in TrtGptModelInflightBatching to reflect the new method signature.
- Adjusted internal logic to maintain functionality without the need for external request data.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Remove disableLookahead method from GptDecoderBatched and iGptDecoderBatched

- Eliminated the disableLookahead method from both GptDecoderBatched and iGptDecoderBatched interfaces to streamline the API.
- Updated TrtGptModelInflightBatching to utilize a new internal method for handling lookahead disabling, improving encapsulation and clarity.
- Adjusted related logic in the setupDecoderStep method to maintain functionality without the removed method.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Unify sampling configuration and batch slot handling in CreateNewDecoderRequests

- Updated the operator() method to utilize std::optional for sampling configurations, improving flexibility in handling requests.
- Introduced fillBatchSlots and fuseSamplingConfigs methods to streamline batch slot filling and sampling configuration fusion.
- Refactored copySequenceLengths to return batch slots, enhancing clarity and separation of concerns in the code.
- Use new methods to disable lookahead in trtGptModelInflightBatching.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #10240](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10240/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7567](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7567/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #10283](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10283/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11847](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11847/) [ run ] triggered by Bot

Funatiq committed:
refactor: Update CreateNewDecoderRequests to handle context requests more effectively

- Filter out finished context requests from the contextRequests vector before calling CreateNewDecoderRequests.
- Modified the operator() method to directly use finishedContextRequests, enhancing clarity.
- Updated the handling of sampling configurations to ensure a valid value is returned, improving robustness.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #13820](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13820/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #16184](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16184/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #16184](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16184/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #12172](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12172/) completed with status: 'FAILURE'
